core/services/synchronization/telemetry_ingress_batch_client.go

```diff
@@ -0,0 +1,301 @@
+	}
+}
+
+// Start connects the wsrpc client to the telemetry ingress server
+func (tc *telemetryIngressBatchClient) Start() error {
+	return tc.StartOnce("TelemetryIngressBatchClient", func() error {
+		privkey, err := tc.getCSAPrivateKey()
+		if err != nil {
+			return err
+		}
+
+		tc.connect(privkey)
+
+		return nil
+	})
+}
+
+// Close disconnects the wsrpc client from the ingress server and waits for all workers to exit
+func (tc *telemetryIngressBatchClient) Close() error {
+	return tc.StopOnce("TelemetryIngressBatchClient", func() error {
+		close(tc.chDone)
+		tc.wgDone.Wait()
+		return nil
+	})
+}
+
+func (tc *telemetryIngressBatchClient) connect(clientPrivKey []byte) {
+	tc.wgDone.Add(1)
+
+	go func() {
+		defer tc.wgDone.Done()
+
+		serverPubKey := keys.FromHex(tc.serverPubKeyHex)
+
+		conn, err := wsrpc.Dial(tc.url.String(), wsrpc.WithTransportCreds(clientPrivKey, serverPubKey))
+		if err != nil {
+			tc.lggr.Errorf("Error connecting to telemetry ingress server: %v", err)
+			return
+		}
+		defer conn.Close()
+
+		// Initialize a new wsrpc client caller
+		// This is used to call RPC methods on the server
+		if tc.telemClient == nil { // only preset for tests
+			tc.telemClient = telemPb.NewTelemClient(conn)
+		}
+
+		// Wait for close
+		<-tc.chDone
+
+	}()
+}
+
+// getCSAPrivateKey gets the client's CSA private key
+func (tc *telemetryIngressBatchClient) getCSAPrivateKey() (privkey []byte, err error) {
+	keys, err := tc.ks.GetAll()
+	if err != nil {
+		return privkey, err
+	}
+	if len(keys) < 1 {
+		return privkey, errors.New("CSA key does not exist")
+	}
+
+	return keys[0].Raw(), nil
+}
+
+// Send directs incoming telmetry messages to the worker responsible for pushing it to
+// the ingress server. If the worker telemetry buffer is full, messages are dropped
+// and a warning is logged.
+func (tc *telemetryIngressBatchClient) Send(payload TelemPayload) {
+	worker := tc.findOrCreateWorker(payload)
+	select {
+	case worker.chTelemetry <- payload:
+		worker.dropMessageCount.Store(0)
+	case <-payload.Ctx.Done():
+		return
+	default:
+		worker.logBufferFullWithExpBackoff(payload)
+	}
+}
+
+// findOrCreateWorker finds a worker by ContractID or creates a new one if none exists
+func (tc *telemetryIngressBatchClient) findOrCreateWorker(payload TelemPayload) *telemetryIngressBatchWorker {
+	tc.workersMutex.Lock()
+	defer tc.workersMutex.Unlock()
+
+	worker, found := tc.workers[payload.ContractID]
+
+	if !found {
+		worker = NewTelemetryIngressBatchWorker(
+			tc.telemMaxBatchSize,
+			tc.telemSendInterval,
+			tc.telemClient,
+			&tc.wgDone,
+			tc.chDone,
+			make(chan TelemPayload, tc.telemBufferSize),
+			payload.ContractID,
+			tc.globalLogger,
+			tc.logging,
+		)
+		worker.Start()
+		tc.workers[payload.ContractID] = worker
+	}
+
+	return worker
+}
+
+// telemetryIngressBatchWorker pushes telemetry in batches to the ingress server via wsrpc.
+// A worker is created per ContractID.
+type telemetryIngressBatchWorker struct {
```

Comment:
For better readability, I would suggest to move worker to its own file.

Reply:
Worker has been moved to its own file